Ship metrics to Graphite directly.

tobz opened the pull request:
Instead of first shipping metrics to statsd, and then onwards to Graphite, we should just ship them directly to Graphite.

As far as I can tell, we were seeing issues where metrics were being shipped from EMR master nodes to the Graphite machine we had running statsd, over UDP, and things were being mysteriously dropped.

We'll be shipping via TCP now, which eliminates at least one variable.  We're also dropping the baggage of statsd defaults such as namespacing metrics by type (like gauge vs counter) and automatic rate calculations, etc.

azafty468 commented:
Looks pretty straight forward.  Question though, are we really calling our yaml file: scripts/collect-hadoop-metrics.yaml.example ?

tobz commented:
It's an example configuration, for people who aren't us (me, really) and are running this out in the wild. :)

azafty468 commented:
Ah, spot on then!

brianhw commented:
Your earlier comment about how this is being run by openedx installations was that UDP meant they shouldn't be affected.  (It would still require them to configure it, nonetheless, which seems weird).  Now that this is TCP, I'm assuming that argument no longer holds.  Is there a way to make this functionality dependent somewhere on it being explicitly configured?   Here or in run-automated-task.sh?

mulby commented on scripts/collect-hadoop-metrics.py:
```diff
@@ -292,10 +293,9 @@ def collect_metrics(hs_address, metric_templates):
-        print "[collect-hadoop-metrics] shipping metrics {} -> {}".format(metric_name, metric_value)
-        statsd_client.gauge(metric_name, metric_value)
-        statsd_client.incr(metric_name, metric_value)
+        stats_client.send(metric_name, metric_value, emission_ts)
```
We should probably wrap all of this new stuff in a try-except so that it doesn't break for users that aren't running a graphite server.

mulby replied:
Optional, but I think we could do something like this?

```python
stats_client.send_list(metrics)
```
It looks like graphitesend will use the current time if no timestamp is specified.

mulby commented on scripts/collect-hadoop-metrics.py:
```diff
@@ -279,10 +280,10 @@ def collect_metrics(hs_address, metric_templates):
-    statsd_host = statsd_config.get('host', 'localhost')
-    statsd_port = statsd_config.get('port', 8125)
-    statsd_prefix = statsd_config.get('prefix', 'edx.analytics.emr')
+    graphite_config = output_config.get('graphite', {})
```
If this isn't defined, maybe just return without even attempting to send the metrics over?

That way open source users won't have to wait 2 seconds for the TCP connection to timeout.

tobz commented:
I think this sufficiently addresses the concerns now.

There will still be error messages that get spit out, but it won't exit with 1, and so things should not gum up on the script failing to run.

If the configuration file doesn't exist, it will simply stop running.
